File: macos/Timesdump/Timesdump/Services/TimeParser.swift
import Foundation

struct TimeParser {
    
    struct Configuration {
        let minYear: Int
        let maxYear: Int
        let timeFormat: String
        
        init(minYear: Int = 1990, maxYear: Int = 2050, timeFormat: String = "yyyy-MM-dd HH:mm:ss") {
            self.minYear = minYear
            self.maxYear = maxYear
            self.timeFormat = timeFormat
        }
    }
    
    private let configuration: Configuration
    
    init(configuration: Configuration = Configuration()) {
        self.configuration = configuration
    }
    
    /// Parses clipboard content and returns a TimestampResult if valid
    /// - Parameter content: The raw clipboard content
    /// - Returns: A TimestampResult if the content is a valid timestamp, nil otherwise
    func parse(_ content: String) -> TimestampResult? {
        // Step 1: Trim whitespace
        let trimmed = content.trimmingCharacters(in: .whitespacesAndNewlines)
        
        // Step 2: Validate - must contain only digits
        guard !trimmed.isEmpty, trimmed.allSatisfy({ $0.isNumber }) else {
            return nil
        }
        
        // Step 3: Parse as number
        guard let numericValue = Double(trimmed) else {
            return nil
        }
        
        // Step 4: Determine unit (seconds vs milliseconds)
        let isMilliseconds = trimmed.count > 10
        let timestamp: TimeInterval = isMilliseconds ? numericValue / 1000.0 : numericValue
        
        // Step 5: Create date and validate year range
        let date = Date(timeIntervalSince1970: timestamp)
        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)
        
        guard year >= configuration.minYear && year <= configuration.maxYear else {
            return nil
        }
        
        // Step 6: Return result
        return TimestampResult(
            originalValue: trimmed,
            timestamp: timestamp,
            date: date,
            format: configuration.timeFormat,
            isMilliseconds: isMilliseconds
        )
    }
}

File: windows/Timesdump/Services/TimeParser.cs
using Timesdump.Models;

namespace Timesdump.Services;

public class TimeParser
{
    private readonly int _minYear;
    private readonly int _maxYear;
    private readonly string _timeFormat;

    public TimeParser(int minYear = 1990, int maxYear = 2050, string timeFormat = "yyyy-MM-dd HH:mm:ss")
    {
        _minYear = minYear;
        _maxYear = maxYear;
        _timeFormat = timeFormat;
    }

    /// <summary>
    /// Parses clipboard content and returns a TimestampResult if valid
    /// </summary>
    /// <param name="content">The raw clipboard content</param>
    /// <returns>A TimestampResult if the content is a valid timestamp, null otherwise</returns>
    public TimestampResult? Parse(string content)
    {
        // Step 1: Trim whitespace
        var trimmed = content.Trim();

        // Step 2: Validate - must contain only digits
        if (string.IsNullOrEmpty(trimmed) || !trimmed.All(char.IsDigit))
        {
            return null;
        }

        // Step 3: Parse as number
        if (!double.TryParse(trimmed, out var numericValue))
        {
            return null;
        }

        // Step 4: Determine unit (seconds vs milliseconds)
        var isMilliseconds = trimmed.Length > 10;
        var timestamp = isMilliseconds ? numericValue / 1000.0 : numericValue;

        // Step 5: Create date and validate year range
        var date = DateTimeOffset.FromUnixTimeSeconds((long)timestamp).DateTime;
        var year = date.Year;

        if (year < _minYear || year > _maxYear)
        {
            return null;
        }

        // Step 6: Return result
        return new TimestampResult(
            trimmed,
            timestamp,
            date,
            _timeFormat,
            isMilliseconds
        );
    }
}
